src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,805 @@
+                _completionService = _document?.GetLanguageService<CompletionService>();
+                _completionRules = _completionService?.GetRules(CompletionOptions.From(_document!.Project)) ?? CompletionRules.Default;
+
+                // Let us make the completion Helper used for non-Roslyn items case-sensitive.
+                // We can change this if get requests from partner teams.
+                _completionHelper = _document != null ? CompletionHelper.GetHelper(_document) : new CompletionHelper(isCaseSensitive: true);
+                _filterMethod = _completionService == null
+                    ? ((itemsWithPatternMatches, text) => CompletionService.FilterItems(_completionHelper, itemsWithPatternMatches, text))
+                    : ((itemsWithPatternMatches, text) => _completionService.FilterItems(_document!, itemsWithPatternMatches, text));
+
+                // Nothing to highlight if user hasn't typed anything yet.
+                _highlightMatchingPortions = _filterText.Length > 0
+                        && globalOptions.GetOption(CompletionViewOptions.HighlightMatchingPortionsOfCompletionListItems, _document?.Project.Language);
+
+                _showCompletionItemFilters = globalOptions.GetOption(CompletionViewOptions.ShowCompletionItemFilters, _document?.Project.Language);
+            }
+
+            public FilteredCompletionModel? UpdateCompletionList(CancellationToken cancellationToken)
+            {
+                if (ShouldDismissCompletionListImmediately())
+                    return null;
+
+                // Use a dedicated pool to minimize potentially repeated large allocations,
+                // since the completion list could be long with import completion enabled.
+                var itemsToBeIncluded = s_listOfMatchResultPool.Allocate();
+                try
+                {
+                    // Determine the list of items to be included in the completion list.
+                    // This is computed based on the filter text as well as the current
+                    // selection of filters and expander.
+                    AddCompletionItems(itemsToBeIncluded, cancellationToken);
+
+                    // Decide if we want to dismiss an empty completion list based on CompletionRules and filter usage.
+                    if (itemsToBeIncluded.Count == 0)
+                        return HandleAllItemsFilteredOut();
+
+                    // Decide the item to be selected for this completion session.
+                    // The selection is mostly based on how well the item matches with the filter text, but we also need to
+                    // take into consideration for things like CompletionTrigger, MatchPriority, MRU, etc. 
+                    var initialSelection = InitialTriggerReason == CompletionTriggerReason.Backspace || InitialTriggerReason == CompletionTriggerReason.Deletion
```

Comment:
checking for Backspace and Delete seems weird to me given that it looked like your earlier helpers collapsed things down so we only see one of these.

Reply:
Previously we have mixed usage of Roslyn's `CompeltionTriggerKind` and Editor's `CompletionTriggerReason` in the ItemManager, which is confusing. So I changed to use Editor's `CompletionTriggerReason` only in async completion layer, unless we are trying to interact with APIs in Roslyn's completion system (i.e. In Features layer)